src/24_Large_Datasets.md

```diff
@@ -0,0 +1,249 @@
+
+## AVRO
+
+![](images/avro_logo.png)
+
+- Free and open-source raw-oriented data storage format
+- Not supported natively by pandas
+- Relies on schemas
+
+## Zarr
+
+![](images/zarr_logo.png)
+
+- Free and open-source
+- Array oriented file format
+- File storage format for chunked, compressed, N-dimensional arrays
+- Same performances than HDF5
+- More flexible because chunking can be done along any dimension
+
+
+## Comparison 
+
+|     | CSV | HDF5 | Parquet | Feather | Avro | Zarr |
+|-----|-----|------|---------|---------|------|------|
+| Format | Row | Array  | Column | Column | Row | Array |
+| Writing | - - - | - - | + | +++ | ++ | + |
+| File size | - - | - - - |  ++ | + | ++ | ++ | 
+| Compression | no | no | +++ | + | + | +++ | 
+| Reading | - - | - - - | ++ | +++ | +  | ++ |
+
+## Dedicated format
+
+- Best performances
+- Optimized for one specific application or purpose
+- For instance: 
+  - TFRecord format
+  - Cloud Optimized GeoTIFF: Optimze satellite images in cloud storage
+
+
+## Choice of the file format
+
+The optimum format depends on:  
+  
+- the type of data stored: same types or not
+- the type of access: partial access or global access
+- the usage: compatibility with the library or the application used
+
+
+# Where to store large datasets ?
+
+## Distributed data storage
+
+- HDFS
+- Amazon S3
+- Azure storage
+- Google Cloud Storage
+
+# Pandas and large datasets
+
+## Good pratices: Load less data
+
+:::::::::::::: {.columns}
+::: {.column width="50%"}
+
+- Load only required columns for our processing
+
+:::
+::: {.column width="50%"}
+
+```python
+pd.read_parquet("timeseries_wide.parquet", columns=required_columns)
+```
+
+:::
+::::::::::::::
+
+## Good pratices: Use efficient datatypes
+
+:::::::::::::: {.columns}
+::: {.column width="50%"}
+
+- Objective: Reduce memory foot-print
+- Text data column is not memory efficient
+- Use pandas.Categorical for categorical data
+- Use pandas.numeric() to downcast the numeric columns to their smallest types 
+
+:::
+::: {.column width="50%"}
+
+```python
+ts2["name"] = ts2["name"].astype("category")
+ts2["id"] = pd.to_numeric(ts2["id"], downcast="unsigned")
+```
+
+:::
+::::::::::::::
+
+## Good pratices: Use chunking
+
+:::::::::::::: {.columns}
+::: {.column width="50%"}
+
+- Split a large problem into a bunch of small problems. 
+- Each chunk must fits in memory
+
+:::
+::: {.column width="50%"}
+
+```python
+from more_itertools import sliced
+CHUNK_SIZE = 5
+
+index_slices = sliced(range(len(df)), CHUNK_SIZE)
+
+for index_slice in index_slices:
+    chunk = df.iloc[index_slice] # your dataframe chunk ready for use
+```
+
+:::
+::::::::::::::
+
+## Good pratices: Dask
+
+- Dask includes dask.dataframe, a pandas-like API for working with larger than memory datasets in parallel. 
+- Dask can use multiple threads or processes on a single machine, or a cluster of machines to process data in parallel.
+- More information in the next course
+
```

Comment:
Use it only if needed, Pandas is often the best for dataset that fits in memory

Reply:
Ok